### PR TITLE
Handle trailing newline in Jupyter notebook JSON string

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

Handle trailing newline in Jupyter Notebook JSON string similar to how `black`
does it.

## Test Plan

Add test cases when the JSON string for notebook ends with and without a newline.

resolves: #5190

### CODE REVIEW

**Comment by dhruvmanila** [comment]
Current dependencies on/for this PR:
* main
  * **PR #5202** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5202" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/5202?utm_source=stack-comment).

File: crates/ruff/src/jupyter/notebook.rs [hunk 1]
```diff
@@ -576,4 +608,18 @@ print("after empty cells")
...
         Ok(())
     }
+
+    #[test_case(Path::new("before_fix.ipynb"), true; "trailing_newline")]
+    #[test_case(Path::new("no_trailing_newline.ipynb"), false; "no_trailing_newline")]
+    fn test_trailing_newline(path: &Path, trailing_newline: bool) -> Result<()> {
+        let notebook = read_jupyter_notebook(path)?;
+        assert_eq!(notebook.trailing_newline, trailing_newline);
+
+        let mut writer = BufWriter::new(Vec::new());
```
Comment (konstin) [inline crates/ruff/src/jupyter/notebook.rs:618]: nit: no need to wrap a vec into a bufwriter

**Review by konstin — APPROVED** [review]

File: crates/ruff/src/jupyter/notebook.rs [hunk 2]
```diff
@@ -113,20 +113,36 @@ pub struct Notebook {
...
                 IOError {
                     message: format!("{err}"),
                 },
                 TextRange::default(),
             )
         })?);
+        let trailing_newline = reader.seek(SeekFrom::End(-1)).is_ok_and(|_| {
+            let mut buf = [0; 1];
+            reader.read_exact(&mut buf).is_ok_and(|_| buf[0] == b'\n')
+        });
```
Comment (MichaReiser) [inline crates/ruff/src/jupyter/notebook.rs:137]: Do jupyter notebooks support `\r` or `\r\n` line endings?

**Review by MichaReiser — APPROVED** [review]

File: crates/ruff/src/jupyter/notebook.rs (hunk 2 above)
Comment (dhruvmanila) [reply crates/ruff/src/jupyter/notebook.rs:137]: No they don't. At the core, the common package `nbformat` is used which uses `\n` as the default line ending. It adds the newline unconditionally (https://github.com/jupyter/nbformat/blob/57817204230a8f7173b8bec380e571b5d410de0d/nbformat/__init__.py#L196-L208) so we could avoid doing the check while reading but I would prefer to use the `black` implementation which performs this check.

**Pull request merged**